ops/terraform/modules/resources/disk_usage_alarms/main.tf

```
@@ -0,0 +1,192 @@
+locals {
+  account_id = data.aws_caller_identity.current.account_id
+
+  kms_key_arn = data.aws_kms_key.mgmt_cmk.arn
+  kms_key_id  = data.aws_kms_key.mgmt_cmk.key_id
+
+  topic_name = "bfd-server-${var.env}-instance-launch-terminate"
+
+  lambda_timeout_seconds = 30
+  lambda_name            = "manage-disk-usage-alarms"
+
+  alarms_prefix = "bfd-server-${var.env}-alert-disk-usage-percent"
+}
+
+# TODO: This SNS topic (and related resources below) mostly duplicates a similar topic used for
+# server-load; consolidate these in the future
+resource "aws_sns_topic" "this" {
+  name              = local.topic_name
+  kms_master_key_id = local.kms_key_id
+}
+
+resource "aws_autoscaling_notification" "this" {
+  topic_arn = aws_sns_topic.this.arn
+
+  group_names = [
+    data.aws_autoscaling_group.asg.name,
+  ]
+
+  notifications = [
+    "autoscaling:EC2_INSTANCE_LAUNCH",
+    "autoscaling:EC2_INSTANCE_TERMINATE"
+  ]
+}
+
+resource "aws_lambda_permission" "this" {
+  statement_id  = "AllowExecutionFromSNS"
+  action        = "lambda:InvokeFunction"
+  function_name = aws_lambda_function.this.function_name
+  principal     = "sns.amazonaws.com"
+  source_arn    = aws_sns_topic.this.arn
+}
+
+resource "aws_sns_topic_subscription" "this" {
+  topic_arn = aws_sns_topic.this.arn
+  protocol  = "lambda"
+  endpoint  = aws_lambda_function.this.arn
+}
+
+resource "aws_iam_policy" "logs" {
+  name        = "bfd-${var.env}-${local.lambda_name}-logs"
+  description = "Permissions to create and write to bfd-${var.env}-${local.lambda_name} logs"
+  policy      = <<-EOF
+{
+  "Version": "2012-10-17",
+  "Statement": [
+    {
+      "Effect": "Allow",
+      "Action": "logs:CreateLogGroup",
+      "Resource": "arn:aws:logs:us-east-1:${local.account_id}:*"
+    },
+    {
+      "Effect": "Allow",
+      "Action": ["logs:CreateLogStream", "logs:PutLogEvents"],
+      "Resource": [
+        "arn:aws:logs:us-east-1:${local.account_id}:log-group:/aws/lambda/bfd-${var.env}-${local.lambda_name}:*"
+      ]
+    }
+  ]
+}
+EOF
+}
+
+resource "aws_iam_policy" "kms" {
+  name        = "bfd-${var.env}-${local.lambda_name}-kms"
+  description = "Permissions to decrypt mgmt KMS key"
+  policy      = <<-EOF
+{
+  "Version": "2012-10-17",
+  "Statement": [
+    {
+      "Effect": "Allow",
+      "Action": ["kms:Decrypt"],
+      "Resource": ["${local.kms_key_arn}"]
+    }
+  ]
+}
+EOF
+}
+
+resource "aws_iam_policy" "autoscaling" {
+  name        = "bfd-${var.env}-${local.lambda_name}-autoscaling"
+  description = "Permissions for bfd-${var.env}-${local.lambda_name} to describe ASGs"
+  # Unfortunately AWS does not support anything but wildcarding for the resource defition for the
+  # DescribeAutoScalingGroups action
```

As far as the copy is concerned, there's a minor typographical error:
```suggestion
  # Unfortunately AWS does not support anything but wildcarding for the resource definition for the
  # DescribeAutoScalingGroups action
```
Regarding the content, I trust that this doesn't support `Condition` elements, either?

---

> Regarding the content, I trust that this doesn't support `Condition` elements, either?

I hadn't thought of using this, so I gave it a try by adding a `Condition` to restrict the `Describe` action to only ASGs with the `Name` tag corresponding to `bfd-${env}-fhir` (using both `autoscaling:ResourceTag/Name` and `aws:ResourceTag/Name`). Unfortunately, when trying to run the Lambda with the conditional policy, the Lambda failed due to lacking permission each time. Therefore, I don't think it is possible to apply a `Condition` to `autoscaling:Describe*` actions.

For what it's worth, the only thing I can think of applying a `Condition` on is the `Name` tag so that we would only be able to `Describe` the corresponding environment's AutoScaling Group. Is there another `Condition` you had in mind to restrict this further?

Also, there is some conflicting documentation on this with respect to `autoscaling:Describe*` actions (see [this examples page](https://docs.aws.amazon.com/autoscaling/ec2/userguide/security_iam_id-based-policy-examples.html) which lists _multiple_ examples stating that "you must specify them [`Describe*` actions] in a separate statement **without conditions**", but [this page](https://docs.aws.amazon.com/autoscaling/ec2/userguide/control-access-using-iam.html#policy-auto-scaling-condition-keys) shows that `autoscaling:*` actions _do_ support `Conditions` on `ResourceTag`s and the like) so I don't exactly know if the restrictions on `Condition` for `autoscaling:Describe*` is specific to certain conditions like `aws:ResourceTag`/`autoscaling:ResourceTag` or `Condition` as a whole.

---

Seems like this is a bit of a wart. Thanks for taking a closer look. I also appreciate the in-line comment calling it out. Let's just leave it alone.